### PR TITLE
`Programming exercises`: Fix missing grading criteria when starting assessment via scores page

### LINKED ISSUE #5579: `Programming exercises`: Assessment instructions only visible after page refresh when starting assessment from ‘Scores’ list
### Describe the bug

When starting a new manual assessment from the ‘Submissions’ page for a programming exercise, the assessment instructions are not shown in the assessment view.

After a page reload, they are shown.

### To Reproduce

Required: a programming exercise with manual assessment and a structured grading instruction. At least one student has submitted to the exercise.

1. Go to the ‘Scores’ page for the exercise. (Green button on the exercise details page)
2. Click on ‘Assess Submission’ for a submission.
3. In the assessment view, in the right sidebar below the problem statement, there should be a list of grading instructions to enable the drag-and-drop feature onto feedback.
4. They are missing.
5. After a page reload, they become visible.

### Expected behavior

The drag-and-drop assessment instructions are visible without an additional page refresh.

### Screenshots

_No response_

### What browsers are you seeing the problem on?

Firefox

### Additional context

When saving the submission, going back to the submissions page, and then opening the assessment again, the assessment instructions are loaded without an additional refresh.
They also seem to be loaded without refresh when accessing the assessment page via the assessment dashboard.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by b-fein**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context
Fixes #5579.

~~Blocked: Due to the `GradingCriteria` being a `List` instead of a `Set`, the join-query returns duplicates. This will be refactored in a separate PR first.~~ → #8004 

### Description
Due to some refactoring in the past the grading criteria were no longer sent as part of the request to fetch the exercise when starting the assessment. This PR adds a request parameter so that the client can request the grading criteria when needed.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming exercise
  - 1 submission
  - manual assessment enabled
  - grading criteria (you can add them in the exercise settings)

1. Go to the scores page of the exercise.
2. Use the ‘Start/Open/Override assessment’ button in the rightmost column.
3. The assessment view should show the grading criteria in the box on the right side.

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ProgrammingExerciseRepository.java | 100% changed code | ✅                           |
| code-editor-tutor-assessment-container.component.ts | 82% | ✅              |

@coderabbitai ignore

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java [hunk 1]
```diff
@@ -605,23 +612,43 @@ default ProgrammingExercise findByIdWithTemplateAndSolutionParticipationTeamAssi
...
+            }
+        }
+        else {
+            if (withSubmissionResults) {
+                programmingExercise = findWithTemplateAndSolutionParticipationSubmissionsAndResultsAndAuxiliaryRepositoriesById(exerciseId);
+            }
+            else {
+                programmingExercise = findWithTemplateAndSolutionParticipationAndAuxiliaryRepositoriesById(exerciseId);
+            }
+        }
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:643]: Optional / maybe followup:

This is now a bit nested and we have 4 different JPA-Queries that are (only?) used for this purpose. A CiteriaBuilder for this purpose would be a great refactoring. See https://docs.artemis.cit.tum.de/dev/guidelines/criteria-builder.html and / or #6043.

**Review by konrad2002 — APPROVED** [review]
manually tested on legacy ts1, works as expected

**Review by LenaKahle — APPROVED** [review]
Approve, tested on legacy ts11.

**Review by Gusti2010 — APPROVED** [review]
Manually tested on ts2_Legacy. The assessment view shows the assessment criteria in the box on the right side. Works as expected!

**Review by chrisknedl — DISMISSED** [review]
Code looks good.

**Review by AntonGeTUM — COMMENTED** [review]
Tested on legacy ts1: the grading criteria are shown in the box. However, the first time you enter an assessment, the criteria do not appear until you refresh the page. Not sure if it is an issue.

**Review by georgyia — APPROVED** [review]
manually tested on ts1, lgtm

**Review by chrisknedl — APPROVED** [review]
Tested locally. Works as expected now!

**Comment by b-fein** [comment]
Minimal merge conflict as can be seen in 6885a04df709b79b57df2cec25b3735ab2167378.
No additional re-testing required.

**Review by Predixx — DISMISSED** [review]
tested on ts1. LGTM

**Review by chrisknedl — DISMISSED** [review]
Re-approve after merge.

**Review by nata-81 — APPROVED** [review]
Tested on ts1 works as expected.

**Review by chrisknedl — APPROVED** [review]
Reapprove.

**Review by reschandreas — APPROVED** [review]
tested locally, works as expected

**Review by R3dError — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java [hunk 2]
```diff
@@ -35,6 +38,11 @@
 @Repository
 public interface ProgrammingExerciseRepository extends JpaRepository<ProgrammingExercise, Long>, JpaSpecificationExecutor<ProgrammingExercise> {
 
+    default ProgrammingExercise findOneByIdElseThrow(final Specification<ProgrammingExercise> specification, long exerciseId) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:41]: Does this work with Spring Boot 3?

**Pull request merged**